url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/block/BlockInnerTriangularSolver.java

```diff
@@ -0,0 +1,274 @@
+     *
+     */
+    public static void invertLower( double L[] ,
+                                    double L_inv[] ,
+                                    int m ,
+                                    int offsetL ,
+                                    int offsetL_inv )
+    {
+        for( int i = 0; i < m; i++ ) {
+            double L_ii = L[ offsetL + i*m + i ];
+            for( int j = 0; j < i; j++ ) {
+                double val = 0;
+                for( int k = j; k < i; k++ ) {
+                    val += L[offsetL + i*m + k] * L_inv[ offsetL_inv + k*m + j ];
+                }
+                L_inv[ offsetL_inv + i*m + j ] = -val / L_ii;
+            }
+            L_inv[ offsetL_inv + i*m + i ] =  1.0 / L_ii;
+        }
+    }
+
+    /**
+     * <p>
+     * Inverts a square lower triangular matrix:  L = L<sup>-1</sup>
+     * </p>
+     *
+     * @param L Lower triangular matrix being inverted. Over written with inverted matrix.  Modified.
+     * @param m The number of rows and columns.
+     * @param offsetL which index does the L matrix start at.
+     *
+     */
+    public static void invertLower( double L[] ,
+                                    int m ,
+                                    int offsetL )
+    {
+        for( int i = 0; i < m; i++ ) {
+            double L_ii = L[ offsetL + i*m + i ];
+            for( int j = 0; j < i; j++ ) {
+                double val = 0;
+                for( int k = j; k < i; k++ ) {
+                    val += L[offsetL + i*m + k] * L[ offsetL + k*m + j ];
+                }
+                L[ offsetL + i*m + j ] = -val / L_ii;
+            }
+            L[ offsetL + i*m + i ] =  1.0 / L_ii;
+        }
+    }
+
+    /**
+     * <p>
+     * Solves for non-singular lower triangular matrices using forward substitution.
+     * <br>
+     * B = L<sup>-1</sup>B<br>
+     * <br>
+     * where B is a (m by n) matrix, L is a lower triangular (m by m) matrix.
+     * </p>
+     *
+     * @param L An m by m non-singular lower triangular matrix. Not modified.
+     * @param b An m by n matrix. Modified.
+     * @param m size of the L matrix
+     * @param n number of columns in the B matrix.
+     * @param strideL number of elements that need to be added to go to the next row in L
+     * @param offsetL initial index in L where the matrix starts
+     * @param offsetB initial index in B where the matrix starts
+     */
+    public static void solveL( double L[] , double []b ,
+                               int m , int n ,
+                               int strideL , int offsetL , int offsetB )
+    {
+        for( int j = 0; j < n; j++ ) {
+            for( int i = 0; i < m; i++ ) {
+                double sum = b[offsetB + i*n+j];
+                for( int k=0; k<i; k++ ) {
+                    sum -= L[offsetL + i*strideL+k]* b[offsetB + k*n+j];
+                }
+                b[offsetB + i*n+j] = sum / L[offsetL + i*strideL+i];
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 126:134, 264:272.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.